lib/rtp_vp9/vp9_payload_descriptor.ex

```diff
@@ -0,0 +1,347 @@
+  def parse_payload_descriptor(raw_payload)
+
+  def parse_payload_descriptor(<<header::binary-size(1), rest::binary()>>)
+      when byte_size(rest) > 0 do
+    <<i::1, _p::1, _l::1, f::1, _bevz::4>> = header
+
+    if i != f do
+      {:error, :malformed_data}
+    else
+      case get_pid(header, rest, %__MODULE__{first_octet: :binary.decode_unsigned(header)}) do
+        {:ok, {descriptor_acc, rest}} ->
+          case get_layer_indices(header, rest, descriptor_acc) do
+            {:ok, {descriptor_acc, rest}} ->
+              case get_pdiffs(header, rest, 0, descriptor_acc) do
+                {:ok, {descriptor_acc, rest}} ->
+                  case get_scalability_structure(header, rest) do
+                    {:ok, {ss, rest}} ->
+                      {:ok, {%{descriptor_acc | scalability_structure: ss}, rest}}
+
+                    _ ->
+                      {:error, :malformed_data}
+                  end
+
+                _ ->
+                  {:error, :malformed_data}
+              end
+
+            _ ->
+              {:error, :malformed_data}
+          end
+
+        _ ->
+          {:error, :malformed_data}
+      end
+    end
+  end
+
+  def parse_payload_descriptor(_), do: {:error, :malformed_data}
+
+  # no picture id (PID)
+  defp get_pid(<<0::1, _::7>>, rest, descriptor_acc) when byte_size(rest) > 0,
+    do: {:ok, {descriptor_acc, rest}}
+
+  # picture id (PID) present
+  defp get_pid(<<1::1, _::7>>, <<pid::binary-size(1), rest::binary()>>, descriptor_acc)
+       when byte_size(rest) > 0 do
+    case pid do
+      <<0::1, _rest_of_pid::7>> ->
+        {:ok, {%{descriptor_acc | picture_id: :binary.decode_unsigned(pid)}, rest}}
+
+      <<1::1, _rest_of_pid::7>> ->
+        <<second_byte::binary-size(1), rest::binary()>> = rest
+        {:ok, {%{descriptor_acc | picture_id: :binary.decode_unsigned(pid <> second_byte)}, rest}}
+
+      _ ->
+        {:error, :malformed_data}
+    end
+  end
+
+  defp get_pid(_, _, _), do: {:error, :malformed_data}
+
+  # no layer indices
+  defp get_layer_indices(<<_i::1, _p::1, 0::1, _::5>>, rest, descriptor_acc)
+       when byte_size(rest) > 0,
+       do: {:ok, {descriptor_acc, rest}}
+
+  # layer indices present
+  defp get_layer_indices(<<_i::1, _p::1, 1::1, f::1, _::4>>, rest, descriptor_acc)
+       when byte_size(rest) > 0 do
+    case f do
+      # TL0PICIDX present
+      0 ->
+        if byte_size(rest) > 2 do
+          <<tid::3, u::1, sid::3, d::1, tl0picidx, rest::binary()>> = rest
+          {:ok, {%{descriptor_acc | tid: tid, u: u, sid: sid, d: d, tl0picidx: tl0picidx}, rest}}
+        else
+          {:error, :malformed_data}
+        end
+
+      # no TL0PICIDX
+      _ ->
+        if byte_size(rest) > 1 do
+          <<tid::3, u::1, sid::3, d::1, rest::binary()>> = rest
+          {:ok, {%{descriptor_acc | tid: tid, u: u, sid: sid, d: d}, rest}}
+        else
+          {:error, :malformed_data}
+        end
+    end
```

Comment:
Again, use `with`. Applies (almost) wherever you have more than one nesting level